mod-source-record-storage-server/src/main/java/org/folio/consumers/AuthorityLinkChunkKafkaHandler.java

```diff
@@ -0,0 +1,334 @@
+    this.kafkaConfig = kafkaConfig;
+
+    producer = createProducer(SRS_BIB_UPDATE_TOPIC, kafkaConfig);
+  }
+
+  @Override
+  public Future<String> handle(KafkaConsumerRecord<String, String> consumerRecord) {
+    return mapToEvent(consumerRecord)
+      .compose(event -> retrieveRecords(event, event.getTenant())
+        .compose(recordCollection -> mapRecordFieldsChanges(event, recordCollection))
+        .compose(recordCollection -> recordService.saveRecords(recordCollection, event.getTenant()))
+        .map(recordsBatchResponse -> mapRecordsToBibUpdateEvents(recordsBatchResponse, event))
+        .compose(marcBibUpdates -> sendEvents(marcBibUpdates, event, consumerRecord))
+      ).recover(th -> {
+          log.error("Failed to handle {} event", SRS_BIB_UPDATE_TOPIC, th);
+          return Future.failedFuture(th);
+        }
+      );
+  }
+
+  private Future<BibAuthorityLinksUpdate> mapToEvent(KafkaConsumerRecord<String, String> consumerRecord) {
+    try {
+      var event = Json.decodeValue(consumerRecord.value(), BibAuthorityLinksUpdate.class);
+      log.info("Decoded {} event for authorityId {}", BibAuthorityLinksUpdate.class, event.getAuthorityId());
+      return Future.succeededFuture(event);
+    } catch (Exception e) {
+      log.error("Failed to decode event with key {}", consumerRecord.key());
+      return Future.failedFuture(e);
+    }
+  }
+
+  private Future<RecordCollection> retrieveRecords(BibAuthorityLinksUpdate bibAuthorityLinksUpdate, String tenantId) {
+    log.trace("Retrieving bibs for authorityId {}", bibAuthorityLinksUpdate.getAuthorityId());
+    var instanceIds = bibAuthorityLinksUpdate.getUpdateTargets().stream()
+      .flatMap(updateTarget -> updateTarget.getLinks().stream()
+        .map(Link::getInstanceId))
+      .distinct()
+      .collect(Collectors.toList());
+
+    var condition = RecordDaoUtil.getExternalIdsCondition(instanceIds, IdType.INSTANCE)
+      .and(RecordDaoUtil.filterRecordByDeleted(false));
+    return recordService.getRecords(condition, RecordType.MARC_BIB, emptyList(), 0, instanceIds.size(), tenantId);
+  }
+
+  private Future<RecordCollection> mapRecordFieldsChanges(BibAuthorityLinksUpdate bibAuthorityLinksUpdate,
+                                                          RecordCollection recordCollection) {
+    log.debug("Retrieved {} records for authorityId {}.",
+      recordCollection.getTotalRecords(), bibAuthorityLinksUpdate.getAuthorityId());
+
+    return getSubfieldProcessorForEvent(bibAuthorityLinksUpdate).map(subfieldProcessor -> {
+      recordCollection.getRecords().forEach(bibRecord -> {
+        var newRecordId = UUID.randomUUID().toString();
+        var instanceId = bibRecord.getExternalIdsHolder().getInstanceId();
+        var parsedRecord = bibRecord.getParsedRecord();
+        var parsedRecordContent = normalize(parsedRecord.getContent());
+        var fieldsArray = parsedRecordContent.getJsonArray(PARSED_RECORD_FIELDS_KEY);
```

Comment:
As I see it's a little complicated to work with JsonObject. Have you considered working with marc4j Record class? It could be easier with it to process fields/subfields. 
Also it's required to recalculate leader based on fields changes. marc4j Writer could help with it by writing Record object